cpp/ql/lib/experimental/semmle/code/cpp/dataflow/ProductFlow.qll

```diff
@@ -352,32 +355,96 @@ module ProductFlow {
         pragma[only_bind_out](succ.getNode().getEnclosingCallable())
     }
 
+    newtype TKind =
+      TInto(DataFlowCall call) {
+        [any(Flow1::PathNode n).getNode(), any(Flow2::PathNode n).getNode()]
+            .(ArgumentNode)
+            .getCall() = call
+      } or
+      TOutOf(DataFlowCall call) {
+        [any(Flow1::PathNode n).getNode(), any(Flow2::PathNode n).getNode()].(OutNode).getCall() =
+          call
+      } or
+      TJump()
+
+    private predicate into1(Flow1::PathNode pred1, Flow1::PathNode succ1, TKind kind) {
+      exists(DataFlowCall call |
+        kind = TInto(call) and
+        pred1.getNode().(ArgumentNode).getCall() = call and
+        succ1.getNode() instanceof ParameterNode
+      )
+    }
+
+    private predicate out1(Flow1::PathNode pred1, Flow1::PathNode succ1, TKind kind) {
+      exists(ReturnKindExt returnKind, DataFlowCall call |
+        kind = TOutOf(call) and
+        succ1.getNode() = returnKind.getAnOutNode(call) and
+        pred1.getNode().(ReturnNodeExt).getKind() = returnKind
+      )
+    }
+
+    private predicate into2(Flow2::PathNode pred1, Flow2::PathNode succ1, TKind kind) {
+      exists(DataFlowCall call |
+        kind = TInto(call) and
+        pred1.getNode().(ArgumentNode).getCall() = call and
+        succ1.getNode() instanceof ParameterNode
+      )
+    }
+
+    private predicate out2(Flow2::PathNode pred1, Flow2::PathNode succ1, TKind kind) {
+      exists(ReturnKindExt returnKind, DataFlowCall call |
+        kind = TOutOf(call) and
+        succ1.getNode() = returnKind.getAnOutNode(call) and
+        pred1.getNode().(ReturnNodeExt).getKind() = returnKind
+      )
+    }
+
     pragma[nomagic]
     private predicate interprocEdge1(
-      Declaration predDecl, Declaration succDecl, Flow1::PathNode pred1, Flow1::PathNode succ1
+      Declaration predDecl, Declaration succDecl, Flow1::PathNode pred1, Flow1::PathNode succ1,
+      TKind kind
     ) {
       Flow1::PathGraph::edges(pred1, succ1) and
       predDecl != succDecl and
       pred1.getNode().getEnclosingCallable() = predDecl and
-      succ1.getNode().getEnclosingCallable() = succDecl
+      succ1.getNode().getEnclosingCallable() = succDecl and
+      (
+        into1(pred1, succ1, kind)
+        or
+        out1(pred1, succ1, kind)
+        or
+        kind = TJump() and
+        not into1(pred1, succ1, _) and
+        not out1(pred1, succ1, _)
```

Comment:
Do I read this correctly as us keeping the situation from before this PR in case the flow does not involve a function call or return? If so, could we still run into sort of similar FPs as the ones we're trying to rule out in this PR?

Reply:
Thanks for the explanation! It does seem we have some performance regressions in DCA with this?

Reply:
Haha, oh shit. Looking at the join-order metric, I think there's a join-order problem somewhere. The performance was certainly fine before https://github.com/github/codeql/pull/12997/commits/0d6fdc674bb92a8013ba889b34cf9c9e89180c0c. I'll take a look 👀.